x-pack/legacy/plugins/siem/public/pages/detection_engine/rules/create/helpers.ts

```diff
@@ -39,16 +39,52 @@ export const getTimeTypeValue = (time: string): { unit: string; value: number }
   return timeObj;
 };
 
+interface RuleFields {
+  anomalyThreshold: unknown;
+  mlJobId: unknown;
+  queryBar: unknown;
+  index: unknown;
+  ruleType: unknown;
+}
+type QueryRuleFields<T> = Omit<T, 'anomalyThreshold' | 'mlJobId'>;
+type MlRuleFields<T> = Omit<T, 'queryBar' | 'index'>;
+
+const isMlFields = <T>(fields: QueryRuleFields<T> | MlRuleFields<T>): fields is MlRuleFields<T> =>
+  has('anomalyThreshold', fields);
+
+export const filterRuleFieldsForType = <T extends RuleFields>(fields: T, type: RuleType) => {
+  if (isMlRule(type)) {
+    const { index, queryBar, ...mlRuleFields } = fields;
+    return mlRuleFields;
+  } else {
+    const { anomalyThreshold, mlJobId, ...queryRuleFields } = fields;
+    return queryRuleFields;
+  }
+};
+
 export const formatDefineStepData = (defineStepData: DefineStepRule): DefineStepRuleJson => {
-  const { queryBar, isNew, ...rest } = defineStepData;
-  const { filters, query, saved_id: savedId } = queryBar;
-  return {
-    ...rest,
-    language: query.language,
-    filters,
-    query: query.query as string,
-    ...(savedId != null && savedId !== '' ? { saved_id: savedId } : {}),
-  };
+  const ruleFields = filterRuleFieldsForType(defineStepData, defineStepData.ruleType);
+
+  if (isMlFields(ruleFields)) {
+    const { anomalyThreshold, mlJobId, isNew, ruleType, ...rest } = ruleFields;
+    return {
+      ...rest,
+      type: ruleType,
+      anomaly_threshold: anomalyThreshold,
+      ml_job_id: mlJobId,
+    };
+  } else {
+    const { queryBar, isNew, ruleType, ...rest } = ruleFields;
+    return {
+      ...rest,
+      type: ruleType,
+      filters: queryBar?.filters,
+      language: queryBar?.query?.language,
+      query: queryBar?.query?.query as string,
+      saved_id: queryBar?.saved_id,
+      ...(ruleType === 'query' && queryBar?.saved_id ? { type: 'saved_query' as RuleType } : {}),
```

Comment:
nit: You don't need the `as RuleType` here. It will work without an `as cast`.

Reply:
That seems like a Typescript bug to me. If you assign this to an intermediate value and then return it, you'll get an error because `type: string` is too loose.

Reply:
Yeah you are correct. Thanks for correcting my comment there. Appreciate it.